keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1619 @@
+plugin and the other plugins. The PotentialNodes field of unallocated
+ResourceClaims with delayed allocation gets updated now if the field doesn't
+match the current list already.
+
+#### Reserve
+
+A node has been chosen for the Pod.
+
+If necessary, the SelectedNode field of ResourceClaims with delayed allocation
+gets set here and the scheduling attempt gets stopped for now. It will be
+retried when the ResourceClaim status changes.
+
+The Pod gets added to the ReservedFor field of its ResourceClaims to ensure that
+no-one else gets to use those.
+
+#### Unreserve
+
+The Pod gets removed from the ReservedFor field because it cannot be scheduled after
+all.
+
+### kubelet
+
+#### Managing resources
+
+kubelet must ensure that resources are ready for use on the node before running
+the first Pod that uses a specific resource instance and make the resource
+available elsewhere again when the last Pod has terminated that uses it. For
+both operations, kubelet calls a resource kubelet plugin as explained in the next
+section.
+
+Pods that are not listed in ReservedFor or where the ResourceClaim doesn't
+exist at all must not be allowed to run. Instead, a suitable event must be
+emitted which explains the problem. Such a situation can occur as part of
+downgrade scenarios.
+
+In addition, kubelet must remove a Pod from ResourceClaim.ReservedFor before
+deleting the Pod. If this was the last Pod on the node that uses the specific
+resource instance, then NodeUnprepareResource (see below) must have been called
+successfully. This ensures that network-attached resource are available again
+for other Pods, including those that might get scheduled to other nodes. It
+also signals that it is safe to deallocate and delete the ResourceClaim.
+
+#### Communication between kubelet and resource kubelet plugin
+
+Resource kubelet plugins are discovered through the [kubelet plugin registration
+mechanism](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/#device-plugin-registration). A
+new "ResourcePlugin" type will be used in the Type field of the
+[PluginInfo](https://pkg.go.dev/k8s.io/kubelet/pkg/apis/pluginregistration/v1#PluginInfo)
+response to distinguish the plugin from device and CSI plugins.
+
+Under the advertised Unix Domain socket the kubelet plugin provides the following
+gRPC interface. It was inspired by
+[CSI](https://github.com/container-storage-interface/spec/blob/master/spec.md),
+with “volume” replaced by “resource” and volume specific parts removed.
+
+##### `NodePrepareResource`
+
+This RPC is called by kubelet when a Pod that wants to use the
+specified resource is scheduled on a node.  The Plugin SHALL assume
+that this RPC will be executed on the node where the resource will be
+used.  The Plugin SHALL return device name and kind for allocated
+device[s].
+
+The Plugin SHALL create or update json file[s] in CDI format for each
+allocated device. These files SHALL be used by runtime to update
```

Comment:
should we use 'allocated resource' here? CDI json file can contain multiple devices, so most probably one file will be generated for one resource.

Reply:
Let's say "SHALL ensure that there are JSON file(s) in CDI format for the allocated resource".

Perhaps they exist already with the right content, then neither "create" nor "update" are necessary. "allocated resource" is indeed better.

Reply:
Perhaps we need to be more descriptive here. For me "SHALL ensure that there are JSON file(s) in CDI format for the allocated resource" sounds like there is json file for the resource already and plugin don't need to either create or update it. What plugin should do if this is not the case is not clear.

Reply:
"ensure" to me is broader that "create and update": it includes that, if necessary.

But feel free to improve this section in your pending PR.

Reply:
I'm ok with your phrasing. Updated my PR.